lxd/api_cluster.go

```diff
@@ -2233,3 +2242,434 @@ func internalClusterRaftNodeDelete(d *Daemon, r *http.Request) response.Response
+	} else if req.Action == "restore" {
+		return restoreClusterMember(d, r)
+	}
+
+	return response.BadRequest(fmt.Errorf("Unknown action %q", req.Action))
+}
+
+func evacuateClusterMember(d *Daemon, r *http.Request) response.Response {
+	var err error
+	var node db.NodeInfo
+
+	nodeName := mux.Vars(r)["name"]
+
+	// Set node status to EVACUATED
+	err = d.cluster.Transaction(func(tx *db.ClusterTx) error {
+		// Get the node.
+		node, err = tx.GetNodeByName(nodeName)
+		if err != nil {
+			return errors.Wrap(err, "Failed to get node by name")
+		}
+
+		if node.State == db.ClusterMemberStatePending {
+			return fmt.Errorf("Cannot evacuate or restore a pending node")
+		}
+
+		// Set node status to EVACUATED to prevent instances from being created.
+		err = tx.UpdateNodeStatus(node.ID, db.ClusterMemberStateEvacuated)
+		if err != nil {
+			return errors.Wrap(err, "Failed to update node status")
+		}
+
+		return nil
+	})
+	if err != nil {
+		return response.SmartError(err)
+	}
+
+	// Do nothing if the node is already evacuated.
+	if node.State == db.ClusterMemberStateEvacuated {
+		return response.SmartError(fmt.Errorf("Node is already evacuated"))
+	}
+
+	reverter := revert.New()
+	defer reverter.Fail()
+
+	// Ensure node is put into its previous state if anything fails.
+	reverter.Add(func() {
+		d.cluster.Transaction(func(tx *db.ClusterTx) error {
+			tx.UpdateNodeStatus(node.ID, db.ClusterMemberStateCreated)
+
+			return nil
+		})
+	})
+
+	// The instances are retrieved in a separate transaction, after the node is in EVACUATED state.
+	var dbInstances []db.Instance
+
+	err = d.cluster.Transaction(func(tx *db.ClusterTx) error {
+		// If evacuating, consider only the instances on the node which needs to be evacuated.
+		dbInstances, err = tx.GetInstances(db.InstanceFilter{Node: nodeName})
+		if err != nil {
+			return errors.Wrap(err, "Failed to get instances")
+		}
+
+		return nil
+	})
+	if err != nil {
+		return response.SmartError(err)
+	}
+
+	instances := make([]instance.Instance, len(dbInstances))
+
+	for i, dbInst := range dbInstances {
+		inst, err := instance.LoadByProjectAndName(d.State(), dbInst.Project, dbInst.Name)
+		if err != nil {
+			return response.SmartError(errors.Wrap(err, "Failed to load instance"))
+		}
+
+		instances[i] = inst
+	}
+
+	var targetNodeName string
+	var targetNode db.NodeInfo
+
+	run := func(op *operations.Operation) error {
+		metadata := make(map[string]interface{})
+
+		// Stop all instances if needed
+		for _, inst := range instances {
+			if inst.IsRunning() {
+				metadata["evacuation_progress"] = fmt.Sprintf("Stopping %s", inst.Name())
+				op.UpdateMetadata(metadata)
+
+				// Be nice
+				timeout := inst.ExpandedConfig()["boot.host_shutdown_timeout"]
+				val, err := strconv.Atoi(timeout)
+				if err != nil {
+					val = 30
+				}
+
+				err = inst.Shutdown(time.Duration(val) * time.Second)
+				if err != nil {
+					return errors.Wrapf(err, "Failed to stop instance %q", inst.Name())
```

Comment:
This should instead attempt a `inst.Stop` at that point. If `inst.Stop` fails, then check if the instance is still running before reporting an error, if it's stopped, just continue.

Reply:
I think inst.Stop does this for us now as I recently added that.

Reply:
@tomponline `inst.Stop` does check if the instance is running but returns an error if it's not. I need to detect that, so I'm going with:
```go
timeout := inst.ExpandedConfig()["boot.host_shutdown_timeout"]
val, err := strconv.Atoi(timeout)
if err != nil {
	val = 30
}

err = inst.Shutdown(time.Duration(val) * time.Second)
if err != nil {
	err = inst.Stop(false)
	if err != nil && inst.IsRunning() {
		return errors.Wrapf(err, "Failed to stop instance %q", inst.Name())
	}
}
```

Reply:
@monstermunchkin to avoid unnecessary calls to the instance's status logic (which can be expensive if done repeatedly for many instances), I would be tempted to replace all occurrences of:

```go
return fmt.Errorf("The instance is already stopped")
```

With a dedicated error var that you can then check as being the cause of the error returned from both `inst.Shutdown` and `inst.Stop`, rather than calling `inst.IsRunning()` again. Effectively getting the `Stop` and `Shutdown` functions to tell you the instance is already stopped.

Reply:
@tomponline that's a good idea. This should be in a seprate PR though. I'll create one.

Reply:
@tomponline I've rebased the PR and added the changes.